internal/provisionertask/machine_fsm.go

```diff
@@ -0,0 +1,157 @@
+//	Starting --> Failed: StartPermanentErr
+//	Starting --> CancellingStart: machineChange(Dead|Dying while in-flight)
+//	CancellingStart --> Stopping: StartSuccess (instance now exists → compensate)
+//	CancellingStart --> Dead: StartErr(any) (no instance to stop)
+//	Running --> Stopping: machineChange(Dead|Dying|Remove)
+//	Stopping --> Dead: StopSuccess (MarkForRemoval done)
+//	Stopping --> Stopping: StopTransientErr / backoff
+//	Stopping --> Failed: StopPermanentErr
+//	Dead --> [*]
+//	Failed --> [*]
+var allowedTransitions = map[MachineState]set.Strings{
+	StatePending: set.NewStrings(
+		string(StateStarting),
+		string(StateDead),
+	),
+	StateStarting: set.NewStrings(
+		string(StateRunning),
+		string(StateFailed),
+		string(StateCancellingStart),
+		string(StateStarting), // idempotent for transient errors.
+	),
+	StateCancellingStart: set.NewStrings(
+		string(StateStopping),
+		string(StateDead),
+	),
+	StateRunning: set.NewStrings(
+		string(StateStopping),
+	),
+	StateStopping: set.NewStrings(
+		string(StateDead),
+		string(StateFailed),
+		string(StateStopping), // idempotent for transient errors.
+	),
+	StateDead: set.NewStrings(
+		string(StateDead), // terminal, idempotent.
+	),
+	StateFailed: set.NewStrings(
+		string(StateFailed), // terminal, idempotent.
+	),
+}
+
+// MachineFSM represents a finite state machine for a single machine.
+// It tracks the machine's current state and enforces valid state transitions.
+type MachineFSM struct {
+	// ID is the machine's unique identifier.
+	ID string
+
+	// State is the current state of the machine in the FSM.
+	State MachineState
+}
+
+// NewMachineFSM creates a new MachineFSM with the given ID.
+// The FSM is initialized in the Pending state.
+func NewMachineFSM(id string) *MachineFSM {
+	return &MachineFSM{
+		ID:    id,
+		State: StatePending,
+	}
+}
+
+// TransitionTo attempts to transition the FSM to the target state.
+// It returns an error if the transition is not allowed.
+//
+// If the target state is the same as the current state, this is a no-op
+// and returns nil (idempotent transitions are allowed for terminal and
+// transient states).
+func (f *MachineFSM) TransitionTo(target MachineState) error {
+	if f.State == target {
+		return nil
+	}
+
+	// Check if transition is allowed.
+	allowed, exists := allowedTransitions[f.State]
+	if !exists {
+		return errors.Annotatef(ErrIllegalTransition,
+			"machine %s: unknown current state %q", f.ID, f.State)
+	}
+
+	if !allowed.Contains(string(target)) {
+		return errors.Annotatef(ErrIllegalTransition,
+			"machine %s: cannot transition from %q to %q", f.ID, f.State, target)
+	}
+
+	f.State = target
+	return nil
+}
+
+// IsTerminal returns true if the given state is a terminal state.
+// Terminal states are Dead and Failed, which represent the end of a
+// machine's lifecycle in the FSM.
+func IsTerminal(state MachineState) bool {
+	return state == StateDead || state == StateFailed
+}
```

Comment:
If you remove the cyclic nature of the the graph, this becomes a property of the states.


```suggestion
func IsTerminal(state MachineState) bool {
	return len(allowedTransitions[state]) == 0
}
```